cni-plugin/linkerd-cni-validation/src/main.rs

```diff
@@ -0,0 +1,177 @@
+                    }
+                });
+            }
+        })
+        .instrument(tracing::info_span!("srv_handler"));
+
+        tokio::select! {
+            _shutdown = shutdown_rx.signaled() => {
+                return Ok(())
+            }
+            _v = handler => {
+                return Ok(())
+            }
+        };
+    })
+    .instrument(tracing::info_span!("validation_server"));
+
+    let client_handle = tokio::spawn(async move {
+        let timeout = std::time::Duration::from_secs(120);
+        if let Err(_) = tokio::time::timeout(timeout, ready_rx).await {
+            return Err::<(), anyhow::Error>(anyhow!(
+                "timed-out (120s) waiting for server to be ready"
+            ));
+        };
+
+        let target_addr = SocketAddr::new(IpAddr::V4(target_host), target_port);
+        let mut stream = {
+            debug!(original_dst = %target_addr, "building validation client");
+            let socket = TcpStream::connect(target_addr)
+                .await
+                .with_context(|| format!("failed to connect: {}", target_addr))?;
+            let peer = socket.peer_addr()?;
+            debug!(original_dst = %peer, "client connected to validation server");
+            socket
+        };
+
+        tokio::select! {
+            is_readable = stream.readable() => {
+                if let Err(e) = is_readable {
+                    anyhow::bail!("cannot read off client socket {}", e);
+                }
+            }
+
+            () = tokio::time::sleep(timeout) => {
+                anyhow::bail!("timed-out (120s) waiting for socket to become readable");
+            }
+        };
+
+        let mut buf = [0u8; 8];
+        stream.read_exact(&mut buf).await?;
+        if &buf == &bytes {
+            return Ok(());
+        } else {
+            anyhow::bail!(
+                "expected client to receive {:?}, got {:?} instead",
+                &buf,
+                &bytes
+            );
+        }
+    })
+    .instrument(tracing::info_span!("validation_client"));
+
+    let result = client_handle.await?;
+    match result {
+        Ok(_) => {
+            info!("passed validation: iptables rules are properly set-up");
+            shutdown_server(shutdown_tx).await;
+        }
+        Err(e) => {
+            error!("failed validation: {}", e);
+            shutdown_server(shutdown_tx).await;
+            // Exit with ERRNO 111, 'Connection refused' if server does not validate
+            // rules within 2 minutes, connection is unsuccessful, or resp &
+            // req bytes don't match.
+            std::process::exit(111);
+        }
+    }
+
+    srv_handle.await??;
+
+    Ok(())
+}
+
+#[tracing::instrument(name = "shutdown", skip(shutdown_tx))]
+async fn shutdown_server(shutdown_tx: drain::Signal) {
+    info!("sending shutdown signal to server");
+    shutdown_tx.drain().await
+}
+
+fn mk_random_bytes() -> anyhow::Result<[u8; 8]> {
```

Comment:
Just a thought, but this might be slightly simpler as a known static response like "REDIRECTION SUCCESSFUL" or something instead of random bytes.  Having a human readable response here might also come in handy if we're ever debugging and connecting to this server manually.